evtimer: Add function to retrieve the current remaining time of an evtimer event

bergzand opened the pull request:
As the title states, this PR adds a function to evtimer to retrieve the time remaining for an evtimer_event_t.

Expected use: Display of collections/objects with a timeout based on an event. For example `_nib_offl_entry_t` (currently route lifetime isn't displayed with `nib route`).

miri64 commented:
Not that I see the use of this function, but what is your use-case specifically?

bergzand commented:
> Not that I see the use of this function, but what is your use-case specifically?

On second thought, the `nib route show` not displaying route lifetimes is a bad case here as that looks more like a mismatch between internal off-link nib entry and  `gnrc_ipv6_nib_ft_t`representation of the route (maybe due to late merging of the route timeouts).

Second case and the trigger for starting some work on this is the rework to evtimer for RPL. [Here](https://github.com/RIOT-OS/RIOT/pull/8188/files#diff-d3eec410743668380168fc80685e211eL299) the lifetime was dropped from the shell command because this lifetime was replaced by an evtimer_event. I think that with this PR, the lifetime can be included again in the output.

miri64 commented:
The problem I have with this function is that it deactivates IRQ for a comparably long time (*O(n)*). This could potentially harm the real-time capabilities of the entire system...

bergzand commented:
> The problem I have with this function is that it deactivates IRQ for a comparably long time (O(n)). This could potentially harm the real-time capabilities of the entire system...

Yeah, I noticed this problem too. The reason I added the irq_disable is to prevent list modifications while the iteration is done (by a event timer that gets fired and removed from the head of the list). 

Would it be sufficient to disable IRQ only while checking the first event and assume that only one event is removed during the iteration? To me that sounds like asking for trouble, but I don't have enough experience with this to say for sure. This would require a bit of reordening of the code.

miri64 commented:
> Yeah, I noticed this problem too. The reason I added the irq_disable is to prevent list modifications while the iteration is done (by a event timer that gets fired and removed from the head of the list).

I know why you did it, I just wanted to point out that it is problematic ;-).

> Would it be sufficient to disable IRQ only while checking the first event and assume that only one event is removed during the iteration? To me that sounds like asking for trouble, but I don't have enough experience with this to say for sure. This would require a bit of reordening of the code.

If multiple events are scheduled for the same time, this assumption already crumbles, so I agree that it is asking for trouble. I don't have any good alternative. Maybe @kaspar030 can help?

kaspar030 commented:
> Maybe @kaspar030 can help?

Not really. Saving relative timestamps was a design decision, fully aware of the "O(n) how much is left?" semantic.

kaspar030 commented:
But you definitely have to disable interrupts while traversing the list!

cgundogan commented:
Off-topic:
```c
void evtimer_print(const evtimer_t *evtimer)
```
also needs IRQ disabling.

bergzand commented:
> But you definitely have to disable interrupts while traversing the list!

Thought so, thanks for confirming.

bergzand commented:
> Off-topic:
> 
> void evtimer_print(const evtimer_t *evtimer)
> 
> also needs IRQ disabling.

I think it also needs a check whether there actually is an `evtimer->event` and not a `NULL` pointer

bergzand commented:
> The problem I have with this function is that it deactivates IRQ for a comparably long time (O(n)). This could potentially harm the real-time capabilities of the entire system...

I've put up a small benchmark for this to get some numbers. Tested on a nucleo-f401 (STM32F401RE, ARM® 32-bit Cortex® -M4 CPU @84Mhz). The evtimer_msg test is adjusted to include a time difference between the output:
```
uint32_t before = xtimer_now_usec();
remaining = evtimer_get_offset(&evtimer,(evtimer_event_t*)&events[NEVENTS - 1]);
uint32_t after = xtimer_now_usec();
```

For N events, always querying for the time to last event:

| Events | time (us) |
| --------:|------------:|
| 10       |     4 |
| 50       |     8 |
| 100     |   13 |
| 500     |   56 |
| 1000   | 101 |

miri64 commented:
Maybe add some `@warning` tag to the documentation, that this function is only supposed to be used with debug/user-interactive code, and specify the reasons.

bergzand commented:
> Maybe add some @warning tag to the documentation, that this function is only supposed to be used with debug/user-interactive code, and specify the reasons.

Not that I'm not agreeing on this point with you, but the evtimer_add function has essentially the same issue here with more complex code within the irq_disable block. Why then not add an `@warning` to the whole module with something like:

```
evtimer operations (both additions/deletions and retrieval) disable interrupts.
Please keep this in mind when trying to use these operations frequently within a short time.
```

bergzand commented:
What I'm trying to say is that if somebody decided to use evtimers to repeatedly schedule a number of evtimer_msg_event_t` with short intervals, it could also harm the real-time capabilities of the system.

miri64 commented:
> What I'm trying to say is that if somebody decided to use evtimers to repeatedly schedule a number of `evtimer_msg_event_t` with short intervals, it could also harm the real-time capabilities of the system.

True... I guess we can say it was a design decision then, that this is just the cas with `evtimer` :-/

kaspar030 commented:
> What I'm trying to say is that if somebody decided to use evtimers to repeatedly schedule a number of evtimer_msg_event_t` with short intervals, it could also harm the real-time capabilities of the system.

True. I guess all of the operations might need a "This is O(n)" warning.

I see it a little more pressing to warn about this for ```evtimer_gettoffset()```, though, as it feels like people would be more inclined to regularly check the state of the timer, expecting the operation to be cheap...

bergzand commented:
Added a warning about complexity and IRQ disabled based on the warning in xtimer.h

kaspar030 commented on sys/evtimer/evtimer.c:
```diff
@@ -153,6 +153,26 @@ void evtimer_del(evtimer_t *evtimer, evtimer_event_t *event)
+        evtimer_event_t *list = evtimer->events;
+        /* Get current offset to first event */
+        offset = _get_offset(&evtimer->timer);
+        while (list->next) {
```
how about ```while ((list = list->next))```? I've used that style in the rest of this file.

bergzand replied:
Fixed

kaspar030 commented:
Apart from the style issue, please squash!

bergzand commented:
Fixed and squashed

bergzand commented:
Okay, I messed up a static-test and rebase in trying to quickly fix a trailing whitespace error -_-

miri64 commented on sys/evtimer/evtimer.c:
```diff
@@ -153,6 +153,25 @@ void evtimer_del(evtimer_t *evtimer, evtimer_event_t *event)
+        evtimer_event_t *list = evtimer->events;
+        /* Get current offset to first event */
+        offset = _get_offset(&evtimer->timer);
+        while ((list = list->next)) {
```
Doesn't this skip the first event now?

bergzand replied:
The first event is supposed to be skipped as the current offset of the first event is retrieved at L163

miri64 replied:
Got it!

kaspar030 commented on sys/include/evtimer.h:
```diff
@@ -98,6 +98,20 @@ void evtimer_add(evtimer_t *evtimer, evtimer_event_t *event);
+ * @param[in] event         An event
+ *
+ * @return  The offset of the event
+ * @return  Zero when the event is not in the list
```
Hm, I don't think this is true. In the current implementation, the offsets get added up until the event is found. If it is not found, all offsets are accumulated. right?

bergzand replied:
You're right :(

Do you have a preference for this behaviour? I could add a simple flag to check if a result is found and change the implementation to the spec. Other option is to change the documentation to "Undefined result when the event is not in the list."

kaspar030 replied:
> I could add a simple flag to check if a result is found and change the implementation to the spec. 

Probably the only way.

> Other option is to change the documentation to "Undefined result when the event is not in the list."

I think this won't work. Imagine the timer triggered after calling ```event_getoffset()```, but before that function disables the IRQs. So *when calling the function*, the timer was in the list. Still the output would be wrong. Thus the result of ```event_getoffset()``` could never be trusted...

bergzand replied:
> I think this won't work. Imagine the timer triggered after calling event_getoffset(), but before that function disables the IRQs. So when calling the function, the timer was in the list. Still the output would be wrong. Thus the result of event_getoffset() could never be trusted...

Ack :(

bergzand commented:
I've reworked the `evtimer_get_offset()` to keep track of a bool for whether the event was found. I've also noticed a bug in the original implementation where the event requested is the first event (doesn't cause a break in the loop, thus invalid result). Therefor I've decided to update the offset for the first event and then iterate over the full list (including the first element) greatly simplifying this function. Intuitively I have some problems with a get function modifying the objects, but without this I would have to unroll the loop for the first event in the list.

kaspar030 commented on sys/evtimer/evtimer.c:
```diff
@@ -153,6 +154,26 @@ void evtimer_del(evtimer_t *evtimer, evtimer_event_t *event)
+    unsigned state = irq_disable();
+
+    evtimer_event_t *list = (evtimer_event_t *)&evtimer->events;
+    _update_head_offset(evtimer);
```
This is indeed troubling...

miri64 commented:
I think to have this issue resolve we need to redesign `evtimer` a bit. Since `event.h` is merged now, we might also be able to base this timer on that.

smlng commented:
any news on this one and how to proceed?

miri64 commented:
> any news on this one and how to proceed?

I think my comment sums it up quite nicely:

> I think to have this issue resolve we need to redesign `evtimer` a bit. Since `event.h` is merged now, we might also be able to base this timer on that.
